File: akkademaid-java/src/main/java/com/gms/askdemo/AskDemoArticleParser.java
package com.gms.askdemo;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;











import com.akkademy.messages.GetRequest;
import com.akkademy.messages.SetRequest;
import com.gms.askdemo.message.ArticleBody;
import com.gms.askdemo.message.HttpResponse;
import com.gms.askdemo.message.ParseArticle;




import com.gms.askdemo.message.ParseHtmlArticle;

import scala.PartialFunction;
import scala.runtime.BoxedUnit;
import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSelection;
import akka.actor.Status;
import akka.japi.pf.ReceiveBuilder;
import akka.util.Timeout;
import static akka.pattern.Patterns.ask;
import static scala.compat.java8.FutureConverters.toJava;

public class AskDemoArticleParser extends AbstractActor {
	
	private final ActorSelection cacheActor;
	private final ActorSelection httpClientActor;
	private final ActorSelection articleParseActor;
	private final Timeout timeout;
	
	public AskDemoArticleParser(String cacheActorPath,
			String httpClientActorPath, String articleParseActorPath,
			Timeout timeout) {
		this.cacheActor = context().system().actorSelection(cacheActorPath);
		this.httpClientActor = context().system().actorSelection(httpClientActorPath);
		this.articleParseActor = context().system().actorSelection(articleParseActorPath);
		this.timeout = timeout;
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	@Override
	public PartialFunction<Object, BoxedUnit> receive() {
		return ReceiveBuilder
				.match(ParseArticle.class, msg -> {
					final CompletionStage cacheResult = toJava(ask(cacheActor, new GetRequest(msg.uri), timeout));
					final CompletionStage result = cacheResult.handle((x, t) -> {
						return (x != null)
								? CompletableFuture.completedFuture(x)
								: toJava(ask(httpClientActor, msg.uri, timeout))
									.thenCompose(rawArticle -> toJava(
											ask(articleParseActor,
													new ParseHtmlArticle(msg.uri, 
															((HttpResponse)rawArticle).rawArticle), timeout))
								);
					}).thenCompose(x -> x);
					
					final ActorRef senderRef = sender();
					result.handle((x,t) -> {
						if(x != null){
							if(x instanceof ArticleBody){
								String body = ((ArticleBody)x).body;
								cacheActor.tell(new SetRequest(msg.uri, body), self());
								senderRef.tell(body, self());
							}else if(x instanceof String){
								senderRef.tell(x, self());
							}
						}else if(x == null)
							senderRef.tell(new Status.Failure((Throwable)t), self());
						return null;
					});
				}).build();
	}
}

File: akkademaid-scala/src/main/scala/com/gms/askdemo/AskDemoArticleParser.scala
package com.gms.askdemo

import akka.util.Timeout
import akka.actor.Actor
import com.gms.askdemo.message.ParseArticle
import com.akkademy.message.GetRequest
import akka.pattern.ask;
import com.gms.askdemo.message.HttpResponse
import com.gms.askdemo.message.ParseHtmlArticle
import scala.concurrent.Future
import scala.util.Success
import com.gms.askdemo.message.ArticleBody
import com.akkademy.message.SetRequest
import scala.util.Failure
import akka.actor.Status

class AskDemoArticleParser(cacheActorPath: String, httpClientActorPath: String,
    articleParserActorPath: String, implicit val timeout:Timeout) extends Actor{
  
  val cacheActor = context.actorSelection(cacheActorPath)
  val httpClientActor = context.actorSelection(httpClientActorPath)
  val articleParserActor = context.actorSelection(articleParserActorPath)
  import scala.concurrent.ExecutionContext.Implicits.global
  
  override def receive: Receive = {
    case ParseArticle(uri) => 
      val senderRef = sender()
      val cacheResult = cacheActor ? GetRequest(uri)
      val result = cacheResult.recoverWith{
        case _:Exception => 
          val fRawResult = httpClientActor ? uri
          fRawResult flatMap {
            case HttpResponse(rawArticle) => 
              articleParserActor ? ParseHtmlArticle(uri, rawArticle)
            case x =>
              Future.failed(new Exception("unknown response"))
          }
      }
      
      result onComplete { 
        case Success(x: String) =>
          println("cached result!")
          senderRef ! x
        case Success(x: ArticleBody) =>
          cacheActor ! SetRequest(uri, x.body)
          senderRef ! x
        case Failure(t) => 
          senderRef ! Status.Failure(t)
        case x =>
          println("unknown message! " + x)
      }
  }
}